src/tools.rs

```diff
@@ -267,3 +267,180 @@ pub fn decode_swdord(data: &[u8]) -> i64 {
+    }
+
+    #[inline]
+    fn ensure(&self, len: usize) -> Option<()> {
+        if self.data.len() < (self.off + len) {
+            return None;
+        }
+        Some(())
+    }
+
+    #[inline]
+    pub fn pos(&self) -> usize {
+        self.off
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.off >= self.data.len()
+    }
+
+    #[inline]
+    pub fn take(&mut self) -> u8 {
+        let b = self.data[self.off];
+        self.off += 1;
+        b
+    }
+
+    #[inline]
+    pub fn peek(&mut self) -> u8 {
+        self.data[self.off]
+    }
+
+    #[inline]
+    pub fn take_slice(&mut self, len: usize) -> Option<&[u8]> {
+        self.ensure(len)?;
+        let s = &self.data[self.off..(self.off + len)];
+        self.off += len;
+        Some(s)
+    }
+
+    #[inline]
+    pub fn decode_leb128_128(&mut self) -> Option<u128> {
+        let mut sz = 0;
+        let mut v: u128 = 0;
+        for c in &self.data[self.off..] {
+            v |= ((c & 0x7f) as u128) << sz;
+            sz += 7;
+            if (c & 0x80) == 0 {
+                self.off += sz / 7;
+                return Some(v);
+            }
+        }
+        None
+    }
+
+    #[inline]
+    pub fn decode_leb128(&mut self) -> Option<u64> {
+        self.decode_leb128_128().map(|v| v as u64)
+    }
+
+    pub fn decode_leb128_128_s(&mut self) -> Option<i128> {
+        let saved_off = self.off;
+        if let Some(v) = self.decode_leb128_128() {
+            let s = self.off - saved_off;
+            let s_mask: u128 = 1 << (s * 7 - 1);
+            return if (v & s_mask) != 0 {
+                // negative
+                Some((v as i128) - (s_mask << 1) as i128)
+            } else {
+                Some(v as i128)
+            };
+        }
+        None
+    }
+
+    pub fn decode_leb128_s(&mut self) -> Option<i64> {
+        self.decode_leb128_128_s().map(|v| v as i64)
+    }
+
+    #[inline(always)]
+    pub fn decode_uhalf(&mut self) -> Option<u16> {
+        self.ensure(2)?;
+        let data = &self.data[self.off..];
+        self.off += 2;
+        Some((data[0] as u16) | ((data[1] as u16) << 8))
+    }
+
+    #[cfg(test)]
+    #[inline(always)]
+    pub fn decode_shalf(&mut self) -> Option<i16> {
+        let uh = self.decode_uhalf()?;
+        Some(if uh >= 0x8000 {
+            ((uh as i32) - 0x10000) as i16
+        } else {
+            uh as i16
+        })
+    }
+
+    #[inline(always)]
+    pub fn decode_uword(&mut self) -> Option<u32> {
+        self.ensure(4)?;
+        let data = &self.data[self.off..];
+        self.off += 4;
+        Some(
+            (data[0] as u32)
+                | ((data[1] as u32) << 8)
+                | ((data[2] as u32) << 16)
+                | ((data[3] as u32) << 24),
+        )
+    }
+
+    #[cfg(test)]
+    #[inline(always)]
+    pub fn decode_sword(&mut self) -> Option<i32> {
```

Comment:
Different standards use different nomenclature. `sword` is arguably pretty uncommon and probably ambiguous. Why not stick closer to the naming that Rust itself uses and go with `decode_i32` instead?

Reply:
It follows the terminologies used by ELF, DWARF, and GSYM.   They share a set of type names (intentional or not).  These names will be easier to match with what is said in specifics.

Reply:
I mean I don't feel strongly, but if the next standard rolls along that does diverge then it's even more confusing. Also, I think it's a fallacy to assume that reader of the code are necessarily familiar with those standards and how they use this terminology. Rust names pretty much forego any random guessing, in addition to being shorter.

Reply:
It is unlikely to happen.  These standards, except GSYM, have a long history.  And the history of these terms is even longer. We should hide these terminologies from the API users.  However, parsers of these formats should as close to what is in the specs as possible to make it easier to comply with standards. And these functions play a translation layer between the standard and Rust worlds.

How do you think moving these functions to a sub-module and implicitly saying these functions are only being used by these formats?

Reply:
> How do you think moving these functions to a sub-module and implicitly saying these functions are only being used by these formats?

It's probably fine to leave them where you have them, as I am not sure an extra module adds anything given the current structure of the code. But up to you.